playground/src/editor.tsx

```diff
@@ -111,6 +117,15 @@ export function Editor(props: {
     setErrors(errList);
   }
 
+  astRef.current = async function updateAst() {
```

Comment:
Everything else looks ok to be copy-pasted, but I think I'd merge `updateAst` and `updateHir` into a single function (`updateAstOrHir`) to cut down on the noise a bit.